components/skus/browser/rs/lib/src/sdk/credentials/fetch.rs

```diff
@@ -63,17 +96,120 @@ where
             return Err(InternalError::OrderUnpaid.into());
         }
 
-        let mut csprng = OsRng;
 
         for item in order.items {
             match item.credential_type {
+                CredentialType::TimeLimitedV2 => {
+                    let mut num_creds: usize = 0;
+                    if let Some(ref metadata) = order.metadata {
+                        let num_intervals = metadata.num_intervals.unwrap_or(0);
+                        let num_per_interval = metadata.num_per_interval.unwrap_or(0);
+                        num_creds = num_intervals * num_per_interval;
+                    }
+
+                    let blinded_creds: Vec<BlindedToken> = match self
+                        .client
+                        .get_time_limited_v2_creds(&item.id)
+                        .await?
+                    {
+                        Some(item_creds) => {
+                            let initialized: bool =
+                                !item_creds.unblinded_creds.unwrap_or_default().is_empty();
+                            let has_existing_creds: bool = !item_creds.creds.is_empty();
+
+                            // are we almost expired
+                            let almost_expired = self
+                                .last_matching_time_limited_v2_credential(&item.id)
+                                .await?
+                                .iter()
+                                .any(|cred| {
+                                    Utc::now().naive_utc()
+                                        > (cred.valid_to - chrono::Duration::days(5))
+                                });
+
+                            event!(
+                                Level::DEBUG,
+                                initialized = initialized,
+                                has_existing_creds = has_existing_creds,
+                                almost_expired = almost_expired,
+                                "debugging states"
+                            );
+
+                            let creds: Vec<BlindedToken> =
+                                item_creds.creds.iter().map(|t| t.blind()).collect();
+
+                            if !initialized && has_existing_creds {
+                                // send existing for signing we have already created this batch
+                                // of creds (failure on first signing request case)
+                                creds
+                            } else if (!has_existing_creds && (!initialized || almost_expired))
+                                || (has_existing_creds && initialized && almost_expired)
```

Comment:
I think this may cause an issue in the following case:
1. credentials are successfully signed and retrieved
2. time passes until almost_expired is true
3. we attempt to refresh, we enter this branch and generate / store new unsigned creds
4. the submit succeeds, however the fetch times out
5. we later re-attempt to refresh, we enter this branch and generate / store new unsigned creds

in step 5, wouldn't we want to re-use the same previously generated creds?

Reply:
https://github.com/brave/brave-core/pull/15465/commits/7c2b2b55b905d85ae7e0c31d60e54459c386af6e addressed in this comment, saving state and referring to explicit state instead of indirect assumption of state.